lib/jerboa/client/credentials.ex

```diff
@@ -1,29 +1,56 @@
 defmodule Jerboa.Client.Credentials do
   @moduledoc false
-  ## Data structure containing client's authentication data
+  ## Data structures containing client's authentication data
 
   defstruct [:username, :realm, :secret, :nonce]
 
-  @type t :: %__MODULE__{
-    username: String.t | nil,
-    realm:    String.t | nil,
-    secret:   String.t | nil,
-    nonce:    String.t | nil
-  }
+  @type t :: Initial.t | Final.t
+
+  defmodule Initial do
+    defstruct [:username, :secret]
+
+    @type t :: %__MODULE__{
+      username: String.t,
+      secret:   String.t
+    }
+  end
+
+  defmodule Final do
+    defstruct [:username, :secret, :realm, :nonce]
+
+    @type t :: %__MODULE__{
+      username: String.t,
+      secret:   String.t,
+      realm:    String.t,
+      nonce:    String.t
+    }
+  end
+
+  @spec initial(String.t, String.t) :: Initial.t
+  def initial(username, secret)
+    when is_binary(username) and is_binary(secret) do
+    %Initial{username: username, secret: secret}
+  end
+
+  @spec finalize(t, String.t, String.t) :: Final.t
+  def finalize(%Initial{} = creds, realm, nonce)
+    when is_binary(realm) and is_binary(nonce) do
+    %Final{
+      username: creds.username,
+      secret: creds.secret,
+      realm: realm,
+      nonce: nonce
+    }
+  end
+  def finalize(%Final{} = creds, _, _), do: creds
```

Comment:
Do we need this clause for convenience? After the split it seems superfluous and by eliminating it we could make the spec narrower, too.